### PR TITLE
Wrap the handler in order to catch panics

### LINKED ISSUE #221: Handler panics hang the runtime
If a handler panics it will get picked up by the executor but it will just hang.

Example:
```rust
type Error = Box<dyn std::error::Error + Send + Sync + 'static>;

#[tokio::main]
async fn main() -> Result<(), Error> {
    lambda::run(lambda::handler_fn(handler)).await?;
    Ok(())
}

#[derive(Debug)]
enum MyError {
    Unknown,
}

impl std::fmt::Display for MyError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            match self {
                MyError::Unknown => "Unknown",
            }
        )
    }
}

impl std::error::Error for MyError {}

async fn handler(_: serde_json::Value) -> Result<serde_json::Value, MyError> {
    panic!("imagine an unwrap or unimplemented! here");
}
```

It's possible to work around this by introducing a shim:
```rust
[tokio::main]
async fn main() -> Result<(), Error> {
    lambda::run(lambda::handler_fn(wrapped_handler)).await?;
    Ok(())
}

async fn wrapped_handler(event: serde_json::Value) -> Result<serde_json::Value, MyError> {
    let handle = tokio::spawn(async { handler(event).await }).await;
    handle.map_err(|_| MyError::Unknown.into())?
}
```

### PR DESCRIPTION
**Pull request by medwards**
Fixes #221

*Description of changes:*
Draft attempt to push panic handling up to the runtime. **Does not compile right now**

By submitting this pull request

- [*] I confirm that my contribution is made under the terms of the Apache 2.0 license.
- [*] I confirm that I've made a best effort attempt to update all relevant documentation.

### CODE REVIEW

File: lambda/src/lib.rs [hunk 1]
```diff
@@ -211,15 +211,27 @@ where
...
+                    },
+                }
+                .into_req()?,
+            },
+            Err(panic) => EventErrorRequest {
                 request_id,
                 diagnostic: Diagnostic {
-                    error_message: format!("{:?}", e),
-                    error_type: type_name_of_val(e).to_owned(),
+                    error_message: "Unexpected panic".to_string(),
```
Comment (softprops) [inline lambda/src/lib.rs:233]: Unless I'm reading this wrong, it looks like we are discarding useful debugging information like the type name of the error and the debug display of the error. Is this intended?

Comment (medwards) [reply lambda/src/lib.rs:233]: You're right, that's the other part of why this is in draft. I wasn't able to figure out how to get something meaningful out of the panic info yet.

**Comment by medwards** [comment]
> Does this prevent any the ability for users to set their own panic hook bevavior? An motivator for enabling that would be to allow for users to export unhandled errors into an error tracking system of their choice

That's an interesting angle to take too. Users who want to track their own panics might be able to wrap them in yet another `tokio::spawn` though I haven't experimented along those lines yet. This wouldn't be that different from what users have to do today anyways so it wouldn't be a heavy imposition.

I'll keep this in mind though and try to provide some experiments along these lines when I get this to compile.

**Comment by bahildebrand** [comment]
Haven't seen any progress on this draft in a while. Closing for now. Feel free to re-open if you make more progress.

**Pull request closed without merging**